content/en/blog/2022/cni-mode/index.md

```diff
@@ -0,0 +1,69 @@
+---
+title: "Merbridge CNI Mode"
+linkTitle: "Merbridge CNI Mode"
+date: 2022-05-18
+weight: 1
+description: This blog explains how CNI works in Merbridge.
+---
+
+<!--Merbridge CNI 模式的出现，旨在能够更好地适配服务网格的功能。之前没有 CNI 模式时，Merbridge 能够做得事情比较有限。其中最大的问题是不能适配注入 Istio 的 Sidecar Annotation，这就导致 Merbridge 无法排除某些端口或 IP 段的流量等。同时，由于之前 Merbridge 只处理 Pod 内部的连接请求，这就导致，如果是外部发送到 Pod 的流量，Merbridge 将无法处理。-->
+The CNI mode is designed to make Merbridge work better with service mesh features. Before having this mode, Merbridge cannot exclude traffic from certain ports or IP ranges, because Istio Annotations configured for a pod cannot take effect in Merbridge. Besides, it can only process connection requests in pods. Traffic from external to pods cannot be processed.
+
+<!--为此，我们精心设计了 Merbridge CNI，旨在解决这些问题。-->
+Therefore, we developed the Merbridge CNI mode to address these issues.
+
+<!--## 为什么需要 CNI 模式？-->
+## Why CNI mode is needed
+<!--其一，之前的 Merbridge 只有一个很小的控制面，其监听 Pod 资源，将当前节点的 IP 信息写入 `local_pod_ips` 的 map，以供 connect 使用。但是，connect 程序由于工作在主机内核层，其无法知道当前正在处理的是哪个 Pod 的流量，就没法处理如 `excludeOutboundPorts` 等配置。为了能够适配注入 `excludeOutboundPorts` 的 Sidecar Annotation，我们需要让 eBPF 程序能够得知当前正在处理哪个 Pod 的请求。-->
+First, Merbridge has a small control plane, which listens to Pods and writes the current node IP into the map of `local_pod_ips` for `connect` to use. However, `connect` works in the host kernel. It needs to know a pod's identity before responding to configurations such as `excludeOutboundPorts`. In order to make Sidecar Annotations injected with `excludeOutboundPorts` effective, we need to let eBPF know the request is from which Pod.
+<!--为此，我们设计了一套方法，与 CNI 配合，能够获取当前 Pod 的 IP，以适配针对 Pod 的特殊配置。-->
+To this end, we designed a method that, in cooperation with CNI, can obtain the current Pod IP to validate special configurations for the Pod.
+<!--其二，在之前的 Merbridge 版本中，只有 connect 会处理主机发起的请求，这在同一台主机上的 Pod 互相通讯时，是没有问题的。但是在不同主机之间通讯时就会出现问题，因为按照之前的逻辑，在跨节点通讯时流量不会被修改，这会导致在接收端还是离不开 iptables。-->
+Second, in older Merbridge versions, only `connect` would process requests from the host, which was alright for same-node pod communication. However, in the case of cross-node pod communication, traffic will not be modified according to the previous logic. It means iptables are still necessary on the receiving end.
+<!--这次，我们依靠 XDP 程序，解决入口流量处理的问题。因为 XDP 程序需要挂载网卡，所以也需要借助 CNI。-->
+Here, we turned to the XDP program for ingress traffic processing. The XDP program needs to mount a network card, also making it necessary to have a CNI mode.
```

Comment:
Here, we turned to the XDP program for processing the ingress traffic. The XDP program needs to mount a network card, also making it necessary to have a CNI mode.